### ISSUE TITLE
negative loss

### ISSUE THREAD

**Issue by huangnengCSU** [HawkAaron/warp-transducer#50]
Hi :
I am using the GPU version RNNTLoss, but the loss is negative.
the shape of logits and targets is as follows:
targets: [batch_size, U]
logits: [batch_size, T, U+1, hidden]
RNNTloss(logits, targets, input_len, target_len)
<img width="1211" alt="屏幕快照 2019-09-24 下午9 04 43" src="https://user-images.githubusercontent.com/28866823/65513980-f685a280-df0e-11e9-8913-2ca16e6a5aff.png">
However when I run on cpu, the loss is normal.
<img width="1248" alt="屏幕快照 2019-09-24 下午9 22 53" src="https://user-images.githubusercontent.com/28866823/65515226-7dd41580-df11-11e9-99d0-f8fc32be7ba0.png">

**Comment by huangnengCSU** [comment]
@HawkAaron 
there is no activation function before rnntloss

**Comment by HawkAaron** [comment]
Did you follow the same procedure as bellow?

GPU: logits -> RNNTLoss
CPU: logits -> log_softmax -> RNNTLoss

**Comment by huangnengCSU** [comment]
@HawkAaron 
my procedure is as bellows.
CPU: logits -> RNNTLoss
GPU: logits -> RNNTLoss
both of them have no activation before RNNTLoss.

**Comment by HawkAaron** [comment]
Could you update the code and rebuild again?  Everything is ok in my environment.
please provide more information if this problem happened again. (python, pytorch, cuda, gcc)

**Comment by huangnengCSU** [comment]
I have updated the latest code. But I haven't solve the problem. So I calculate the RNNTLoss on cpu at last. Maybe I will find the problem later when I debug my code.

**Comment by kate-chuikova** [comment]
I have the same problem when using tensorflow 1.15.0.rc2 (CUDA 10, gcc 7.4.0)
all tests (gpu_test, cpu_tests, tensorflow tests) are passed

Previously I used tf 1.10 and everything was ok, then I upgraded tf, rebuild RNNTLoss, and get negative loss.

I install RNNTLoss using the following steps in docker:

ENV CUDA_HOME /usr/local/cuda
ENV LD_LIBRARY_PATH="/usr/local/cuda/lib64:${LD_LIBRARY_PATH}"
RUN cd /rnnt/rnnt-loss && mkdir /rnnt/rnnt-loss/build
RUN cd /rnnt/rnnt-loss/build && cmake -DCUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME .. && \
    make
RUN ldconfig /usr/local/cuda/lib64/stubs && \
    cd /rnnt/rnnt-loss/tensorflow_binding && CUDA=$CUDA_HOME python3 setup.py install

**Comment by kate-chuikova** [comment]
update:
I tried to rerun experiment with updated version of RNNTLoss and tf 1.10 and got negative loss.
My previous RNNTloss version was in 6e33845c9acecce0c9489f1f2c333f2ef548d986 commit and everything was OK in tf 1.10 (but this version doesn't work with tf 1.15)

Also, I got the following error using tf 1.10:
 src/warprnnt_op.cc:8:52: fatal error: tensorflow/core/framework/bounds_check.h: No such file or directory
this file was renamed for tf >= 1.14 support and doesn't exist now for tf 1.10

**Comment by Sudharsansai** [comment]
Just a quick question: Should the blank label be compulsorily 0? What about GPU vs CPU?

**Comment by digital10111** [comment]
I observed that a high learning rate, led to a steep decline in the loss and subsequently the loss started becoming negative for some examples in the batch. This was observed with a LR of 1e-4, on my data. Unsure, if the cause is the network code or rnnt-loss code. Any comments?

**Comment by BuaaAlban** [comment]
> UPD:
> I tried to rerun experiment with updated version of RNNTLoss and tf 1.10 and got negative loss.
> My previous RNNTloss version was in [6e33845](https://github.com/HawkAaron/warp-transducer/commit/6e33845c9acecce0c9489f1f2c333f2ef548d986) commit and everything was OK in tf 1.10 (but this version doesn't work with tf 1.15)
> 
> Also, I got the following error using tf 1.10:
> src/warprnnt_op.cc:8:52: fatal error: tensorflow/core/framework/bounds_check.h: No such file or directory
> this file was renamed for tf >= 1.14 support and doesn't exist now for tf 1.10
> 
> UPD2:
> I found a bug in my code: the problem was in wrong blank_index, everything is OK now.

I also have the question that is it reasonable to have negative loss, maybe it is because the probability i s the sum of all alignments, which means the probability  can be larger than 1 and the loss can be negative?
BTW, what do you mean by "the problem was in wrong blank_index"??

**Comment by HawkAaron** [comment]
The recent pull request #64 may fix this issue.

**Comment by HawkAaron** [comment]
@Sudharsansai  The blank label can be any class inside your vocab (with blank included).
